core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -6536,11 +6540,49 @@ sealed trait Exit[+E, +A] extends ZIO[Any, E, A] { self =>
     f: (A, B) => C,
     g: (Cause[E], Cause[E1]) => Cause[E1]
   ): Exit[E1, C] =
-    (self, that) match {
-      case (Success(a), Success(b)) => Exit.succeed(f(a, b))
-      case (Failure(l), Failure(r)) => Exit.failCause(g(l, r))
-      case (e @ Failure(_), _)      => e
-      case (_, e @ Failure(_))      => e
+    self match {
+      case Success(l) =>
+        that match {
+          case Success(r)     => Exit.succeed(f(l, r))
+          case f: Failure[E1] => f
+        }
+      case e @ Failure(c1) =>
+        that match {
+          case Failure(c2) => Exit.failCause(g(c1, c2))
+          case _           => e
+        }
+    }
+
+  private def zipLeftWith[E1 >: E](that: Exit[E1, Any])(
```

Comment:
You could avoid specialization of left and right if this was in a companion object, no?

Reply:
Huh, good point, I'll do that